hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/HoodieHiveFileSliceReader.java

```diff
@@ -0,0 +1,190 @@
+import org.apache.hudi.exception.HoodieException;
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.hudi.hadoop.realtime.RealtimeSplit;
+import org.apache.hudi.hadoop.utils.HoodieHiveUtils;
+import org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils;
+import org.apache.hudi.internal.schema.InternalSchema;
+import org.apache.hudi.io.storage.HoodieFileReader;
+
+import org.apache.avro.Schema;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.metastore.api.hive_metastoreConstants;
+import org.apache.hadoop.hive.serde2.ColumnProjectionUtils;
+import org.apache.hadoop.mapred.FileSplit;
+import org.apache.hadoop.mapred.InputSplitWithLocationInfo;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils.getMergedLogRecordScanner;
+import static org.apache.hudi.io.storage.HoodieFileReaderFactory.getReaderFactory;
+
+public abstract class HoodieHiveFileSliceReader implements FileSliceReader {
+
+  private static final Logger LOG = LogManager.getLogger(HoodieHiveFileSliceReader.class);
+
+  private final FileSplit split;
+  private final JobConf jobConf;
+  private final HoodieTableMetaClient metaClient;
+  private final Properties payloadProps = new Properties();
+
+  // table schema from metadata
+  private final Schema schema;
+
+  // other schema handles
+  private Schema readerSchema;
+  private Schema writerSchema;
+
+  private String readMode;
+
+  public HoodieHiveFileSliceReader(InputSplitWithLocationInfo split, JobConf jobConf) {
+    this.split = (FileSplit) split;
+    this.jobConf = jobConf;
+    HoodiePartitionMetadata metadata = null;
+    try {
+      Path path = this.split.getPath();
+      metadata = new HoodiePartitionMetadata(path.getFileSystem(jobConf), path);
+      metadata.readFromFS();
+      String basePath = String.valueOf(HoodieHiveUtils.getNthParent(path, metadata.getPartitionDepth()));
+      this.metaClient = HoodieTableMetaClient.builder().setConf(jobConf).setBasePath(basePath).build();
+      this.payloadProps.setProperty(HoodiePayloadProps.PAYLOAD_ORDERING_FIELD_PROP_KEY, metaClient.getTableConfig().getPreCombineField());
+      this.schema = new TableSchemaResolver(metaClient).getTableAvroSchema();
+    } catch (IOException e) {
+      throw new HoodieIOException("Failed to initialize metaclient for split: " + this.split);
+    } catch (Exception e) {
+      throw new HoodieException("Failed to get table avro schema.");
+    }
+  }
+
+  @Override
+  public FileSliceReader project(InternalSchema schema) {
+    // should we convert to avro Schema?
```

Comment:
Yes, we can just convert it to Avro schema and then use the one you've already impl'd